### ISSUE TITLE
Problem with Malli coercion (was: Using Malli registry for coercion/validation?)

### ISSUE THREAD

**Issue by mvarela** [metosin/reitit#453]
UPDATE: the error seems to happen regardless of whether a custom registry is used or not; see https://github.com/mvarela/coercion-bug2 for a minimal repro.

(apologies, I had opened this under `metosin/malli`, when it rather should be here...)

I have defined a Malli registry with a bunch of schemas, and would like to use it in my routes for the `:parameters` and `:response` keys (I'm using `reitit/http`).

I have collected these schemas into a `schemas.core/registry` var.  

I did 
```
:data {:coercion (reitit.coercion.malli/create {:registry schemas.core/registry}) ...
```
when creating the router, but when I try to use one of the schemas defined in the registry (`::schemas.issues/issue-response`), it complains about that being an invalid schema:

```
Execution error (ExceptionInfo) at reitit.dev.pretty/exception (pretty.cljc:230).
-- Router creation failed ------------------------------------ reitit.http:46 --

:malli.core/invalid-schema {:schema :com.profilence.issue-streams.schemas.issues/issue-response}

{:schema :com.profilence.issue-streams.schemas.issues/issue-response}

-------------------------------------------------------------------------------
```

So I then tried using `m/schema` with my registry in the actual route definition:

```
                       :responses {200 {:body  (m/schema [:map [:project string?
                                                                :version string?
                                                                :issues [:sequential [:map [:issue ::schemas.issues/issue-response]
                                                                                      [:instance-count pos-int?]]]]] {:registry schemas.core/registry})}}
```

This seems to work, but when trying it out, it produces a malformed output (copied from Swagger output, hence the JSON format):
```
{
  "project": "tunkio"
}
```
Whereas the correctly-shaped output would be (obtained by commenting out the `:response` bit in the route definition)
```
{
  "project": "tunkio",
  "version": "14",
  "issues": [
    {
      "issue": {
        "type": "java-crash",
        "exceptiontype": "java.lang.IllegalStateException",
        "process": "xxx.server",
        "project": [
          "tunkio"
        ],
        "version": [
          "14"
        ],
        "canonical-message": "UiAutomation not connected!",
        "issue-id": "9652c728-99b4-5fab-9a54-f121730f7159"
      },
      "instance-count": 28
    },
    {
      "issue": {
        "type": "native-crash",
        "category": 1,
        "process": "xxx.master",
        "project": [
          "tunkio"
        ],
        "version": [
          "14"
        ],
        "canonical-message": "'terminating'",
        "issue-id": "d25a5de7-00b2-5732-ae04-216c47c04e11"
      },
      "instance-count": 6
    }
  ]
}
```

I have tried validating the data I get from my handler against the schema above, and it works:
```
  (malli/validate [:map [:project string?
                         :version string?
                         :issues  [:sequential [:map [:issue ::schemas.issues/issue-response]
                                                [:instance-count pos-int?]]] ]]
                  {:project "tunkio",
                   :version "14",
                   :issues
                   ({:issue
                     {:type :java-crash,
                      :exceptiontype "java.lang.IllegalStateException",
                      :process "xxx.server",
                      :project #{"tunkio"},
                      :version #{"14"},
                      :canonical-message "UiAutomation not connected!",
                      :issue-id "9652c728-99b4-5fab-9a54-f121730f7159"},
                     :instance-count 28}
                    {:issue
                     {:type :native-crash,
                      :category 1,
                      :process "xxx.master",
                      :project #{"tunkio"},
                      :version #{"14"},
                      :canonical-message "'terminating'",
                      :issue-id "d25a5de7-00b2-5732-ae04-216c47c04e11"},
                     :instance-count 6})}
                  {:registry schemas.core/registry})
;; => true
```

So, in summary, I believe the schema itself is correct wrt the data I need to produce, but I can't seem to get it to work with reitit.
What is the correct way of specifying a registry for the malli coercion/validation to use?

Thanks in advance!!

**Comment by mvarela** [comment]
I also tried using setting a mutable default registry, and initializing it with my schemas before starting the app (via Integrant), but I still get an error regarding an invalid schema for one of the keys I've defined.

**Comment by mvarela** [comment]
It seems this is a bit deeper than whether the registry is used or not.
I wrote out the actual spec, just in terms of malli "primitives":
```
[:map
 [:project
  string?
  :version
  string?
  :issues
  [:sequential
   [:map
    [:issue
     [:or
      [:map
       [:type [:enum :java-issue :native-crash :anr]]
       [:exceptiontype string?]
       [:process string?]
       [:project [:set {:min 1} [:string {:min 1}]]]
       [:version [:set {:min 1} [:string {:min 1}]]]
       [:canonical-message string?]
       [:issue-id [:string {:min 1}]]]
      [:map
       [:type [:enum :java-issue :native-crash :anr]]
       [:category number?]
       [:process string?]
       [:project [:set {:min 1} [:string {:min 1}]]]
       [:version [:set {:min 1} [:string {:min 1}]]]
       [:canonical-message string?]
       [:issue-id [:string {:min 1}]]]
      [:map
       [:type [:enum :java-issue :native-crash :anr]]
       [:activity string?]
       [:process string?]
       [:project [:set {:min 1} [:string {:min 1}]]]
       [:version [:set {:min 1} [:string {:min 1}]]]
       [:subject string?]
       [:issue-id [:string {:min 1}]]]]]
    [:instance-count pos-int?]]]]]
```

And passing a well-formed value:
```
{:project "DummyProject-Kafka",
 :version "v1.718734659",
 :issues
 ({:issue
   {:type :native-crash,
    :category 2,
    :process "nativeTester",
    :project #{"DummyProject-Kafka"},
    :version #{"v1.718734659" "v1.284409613"},
    :canonical-message "hi there",
    :issue-id "bad569df-fc6d-5b0d-9823-60635a74b561"},
   :instance-count 1})}
```
It still drops the `:version` and `:issues` keys on coercion:
```
--- :leave :reitit.http.coercion/coerce-response ---

  {:error nil,
   :request {:body #<org.eclipse.jetty.server.HttpInputOverHTTP@6d9dc0ce HttpInputOverHTTP@6d9dc0ce[c=0,q=0,[0]=null,s=STREAM]>,
             :character-encoding nil,
             :content-length nil,
             :content-type nil,
             :form-params {},
             :headers {"accept" "application/json",
                       "accept-encoding" "gzip, deflate, br",
                       "accept-language" "en-US,en;q=0.9",
                       "connection" "keep-alive",
                       "host" "localhost:8890",
                       "referer" "http://localhost:8890/index.html",
                       "sec-fetch-dest" "empty",
                       "sec-fetch-mode" "cors",
                       "sec-fetch-site" "same-origin",
                       "user-agent" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.111 Safari/537.36"},
             :parameters {:query {:project "DummyProject-Kafka",
                                  :version "v1.718734659"}},
             :params {"project" "DummyProject-Kafka",
                      "version" "v1.718734659"},
             :path-params {},
             :protocol "HTTP/1.1",
             :query-params {"project" [1;35m"DummyProject-Kafka",
                            "version" "v1.718734659"},
             :query-string "project=DummyProject-Kafka&version=v1.718734659",
             :remote-addr "0:0:0:0:0:0:0:1",
             :request-method :get,
             :scheme :http,
             :server-name "localhost",
             :server-port 8890,
             :ssl-client-cert nil,
             :uri "/api/v0/projects/issuesByVersion",
             :muuntaja/request nil,
             :muuntaja/response {:charset "utf-8",
                                 :format "application/json",
                                 :raw-format "application/json"}},
   :response {[1;33m:body {:project "DummyProject-Kafka",
                     -:issues ({:instance-count 1,
                                :issue {:canonical-message "hi there",
                                        :category 2,
                                        :issue-id "bad569df-fc6d-5b0d-9823-60635a74b561",
                                        :process "nativeTester",
                                        :project #{"DummyProject-Kafka"},
                                        :type :native-crash,
                                        :version #{"v1.284409613"
                                                   "v1.718734659"}}}),
                     -:version "v1.718734659"},
              :status 200}}
```

So it would seem that, regardless of whether the registry is used (still not sure _how_ it should be used...), the coercion is producing erroneous results.

**Comment by mvarela** [comment]
@ikitommi, I produced a minimal repro here: https://github.com/mvarela/coercion-bug2
The bug happens even with a simple schema with a 2-element map, and it is independent of whether a registry is used.

cc: @miikka

**Comment by miikka** [comment]
Definitely looks like a bug – thanks for the clean repro repo! I'll take a look at this now.

**Comment by miikka** [comment]
@mvarela I don't think it's a bug in reitit/malli after all. I think that the map definition is just missing some `[]` and it should look like this:

```clojure
[:map
  [:project string?]
  [:version string?]]
```

Patch against the repro repo:
```diff
diff --git a/src/coercion/bug.clj b/src/coercion/bug.clj
index 313f572..c3461c1 100644
--- a/src/coercion/bug.clj
+++ b/src/coercion/bug.clj
@@ -35,10 +35,8 @@
              ["/ping" {:get {:summary "ping"
                              :parameters {}
                              :responses {200 {:body [:map
-                                                     [:project
-                                                      string?
-                                                      :version
-                                                      string?]]}}
+                                                     [:project string?]
+                                                     [:version string?]]}}
                              :handler (fn [ctx] response)}}]
 
              ["/ping2" {:get {:summary "ping2"
```

With this change it seems to work for me. Could you try it out?

**Comment by mvarela** [comment]
Sure, I'll have a go in my original code, too...  (I have posted those in the comments above, but they are fairly hairy :) )

**Comment by mvarela** [comment]
@miikka, I've updated the repro repository with actual data / schemas. The data used validates correctly, but is later incorrectly coerced.

https://github.com/mvarela/coercion-bug2

**Comment by miikka** [comment]
@mvarela, `response-schema` has the exactly same problem as above with regards the right nesting of vectors

**Comment by mvarela** [comment]
Holy crap, I'm way dumber than I thought  I was! I kept looking at the hairy stuff, and didn't notice the top level. 
I had tried validating and it worked, so I thought the problem had to be somewhere else...  
Thanks for having a look, and sorry for the bother!
